### PR TITLE
add new structure multistep

### PR DESCRIPTION
**Pull request by Neofox**

### CODE REVIEW

File: structures/auction.js [hunk 1]
```diff
@@ -56,7 +57,12 @@ exports.auction = function (config = null) {
       vehicles: s.optional(s.array(s.string())),
       incoterm: s.optional(s.enums(['EXW', 'CIP', 'FCA', 'DAP', 'DPU', 'CPT', 'DDP', 'FAS', 'CFR', 'FOB', 'CIF'])),
       transport: s.optional(s.array(s.string())),
-      dimension: s.optional(s.array(s.string())),
+      dimension: s.dynamic((v,p) => {
+          if (p.branch.some(e => e.hasOwnProperty('options') && e.options.includes('MULTISTEP'))) {
```
Comment (lvenier) [inline structures/auction.js:61]: would have a done a switch as we will move PKV V1/V2 to it.

File: test/multistep.test.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+      "Test 3",
+      "Test 3",
+      "test3@test.test",
+      "300000003",
+      "2021-03-03T15:30:00.000Z"
+    ], struct.multistep)).toBeTruthy()
+
+  })
+
+})
```
Comment (lvenier) [inline test/multistep.test.js:70]: pls add a test with 2 packages with wrong size in one fo them please.

**Review by lvenier — CHANGES_REQUESTED** [review]
update package version also :)

File: structures/multistep.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
+const s = require('superstruct')
+
+const multistep_package = s.size(s.array(s.string()), 13)
+const multistep_address = s.size(s.array(s.string()), 11)
+
+const Multistep = s.size(s.array(s.string()), 27, Infinity)
+
+module.exports = {
+  multistep_address,
```
Comment (fanshan) [inline structures/multistep.js:9]: Sorry but I don't see the relation between `multistep_package`, `multistep_address` and `Multistep`.

File: test/auction.test.js [hunk 4]
```diff
@@ -8,6 +8,7 @@ const Auctions = [{
   key: "jest-" + Math.round((new Date()).getTime()),
   source: ["DEMO"],
   target: ["DEMOEX"],
+  options: ['MULTISTEP'],
```
Comment (fanshan) [inline test/auction.test.js:11]: This is strange because this don't break other test.

Comment (lvenier) [reply test/auction.test.js:11]: @fanshan: The other "options" are not in superstruct yet. They are still being testeed in shaq.
I was hoping Jerome to add test for PKGV2 and V3 but it goes minimalist :)

File: structures/multistep.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
+const s = require('superstruct')
+
+const MAX_PKG = 50
```
Comment (fanshan) [inline structures/multistep.js:3]: @lvenier Agreed with this constant?

Comment (lvenier) [reply structures/multistep.js:3]: I think that should not be a contact but a default value.
pls check https://github.com/yoctu/rtech-struct/blob/main/structures/auction.js#L13 and we can use a config value "config.shaq.pkgmax"

File: structures/auction.js [hunk 6]
```diff
@@ -56,7 +57,19 @@ exports.auction = function (config = null) {
...
+      dimension: s.dynamic((v,p) => {
+          const options = p.branch.map(e => e.hasOwnProperty('options') ? e.options: [])[0]
+          switch (true) {
+              case options.includes('MULTISTEP'):
+                  return multistep
+              case options.includes('PKG_V1'):
+                  return packageV1
+              case options.includes('PKG_V2'):
+                  return packageV2
+              default:
```
Comment (lvenier) [inline structures/auction.js:69]: default should be packageV1

**Review by lvenier — CHANGES_REQUESTED** [review]

File: structures/multistep.js (hunk 5 above)
Comment (fanshan) [reply structures/multistep.js:3]: You're right.

File: structures/multistep.js [hunk 7]
```diff
@@ -0,0 +1,17 @@
+const s = require('superstruct')
+
+const Multistep = function (config = null) {
+  let MaxSteps = 50
```
Comment (fanshan) [inline structures/multistep.js:4]: @Neofox Is possible to test this feature?

File: test/multistep.test.js [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+      "2", "100", "100", "100", "10", "", "no", "", "", "0",
+      "EG2103CF36-A", "35 RUE DE CALAIS", "67100", "STRASBOURG", "France", "FR", "48.5252,7.7824", "Europe/Paris", "test1", "test1", "test1@test.fr", "090809080908", "2021-03-01T01:00:00", "2021-03-01T01:00:00",
+      "EG2103CF36-B", "14 RUE GORGE DE LOUP", "69009", "LYON", "France", "FR", "45.77,4.8041", "Europe/Paris", "test2", "test2", "test2@test.fr", "090809080908", "2021-03-01T01:00:00", "2021-03-01T01:00:00",
+      "2", "100", "100", "100", "10", "", "no", "", "", "0",
+      "EG2103CF36-A", "35 RUE DE CALAIS", "67100", "STRASBOURG", "France", "FR", "48.5252,7.7824", "Europe/Paris", "test1", "test1", "test1@test.fr", "090809080908", "2021-03-01T01:00:00", "2021-03-01T01:00:00",
+      "EG2103CF36-B", "14 RUE GORGE DE LOUP", "69009", "LYON", "France", "FR", "45.77,4.8041", "Europe/Paris", "test2", "test2", "test2@test.fr", "090809080908", "2021-03-01T01:00:00","2021-03-01T01:00:00"
+    ], multistep({shaq: {maxsteps: 2}}))).toBeTruthy()
+  })
+
+  test('Failed: Mulstistep structure (missing a package field)', () => {
```
Comment (lvenier) [inline test/multistep.test.js:46]: Mulstistep ? :see it a couple of times )

File: structures/multistep.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+  if (config && config.shaq && config.shaq.maxsteps) MaxSteps = config.shaq.maxsteps
+
+  return s.size(s.array(s.string()), 38, 38*MaxSteps)
+}
+
+const PackageV1 = s.size(s.array(s.string()), 6)
```
Comment (lvenier) [inline structures/multistep.js:10]: s.size(s.array(s.string()), 6) -> s.size(s.array(s.size(s.string(), 64)), 6) ? 

would be better to have s.tuple : string in first version but could be change later on.

**Review by lvenier — APPROVED** [review]

**Pull request merged**